In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from scipy.stats import spearmanr
from scipy.stats import pearsonr

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/diamonds/diamonds.csv


The goal of this Python script is to find out if there is a relationship between the two variables the carat of a diamond and the price of a diamond, and to see what type of relationship exists between the two variables.

Hypothesis for carat vs price:

I believe that there will be a positive relationship between the carat of a diamond and the price of a diamond, since in general we normally see the more carats a diamond has the higher of a price value the diamond tends to have as well.

To start I will perform a linear regression on the two variables carat of a diamond and price of a diamond. The goal of a regression is to investigate/predict if there is a relationship between the independent and dependent variable. A linear regression specifically is when data is modelled in a straight line in order to predict the value of a variable.

In [2]:
data = pd.read_csv("../input/diamonds/diamonds.csv")
data.head()

Unnamed: 0  carat      cut color clarity  depth  table  price     x     y  \
0           1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98   
1           2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84   
2           3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07   
3           4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23   
4           5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35   

      z  
0  2.43  
1  2.31  
2  2.31  
3  2.63  
4  2.75

To start we attribute arrays x and y to the columns we plan to analyze within this data set. We then check each column to confirm there are no values of 'nan' within the data set using two for loops. Finally we recieve our length of the data set which is 53,940 different values which meets the requirement of 10,000 values.

In [3]:
x = data['carat'].values
y = data['price'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) != 'nan':
        new_y.append(item)
        
print('The length of this dataset is:', len(new_x))

The length of this dataset is: 53940


Now in order to find r squared for our linear regression model, we must first start by finding out the equation for ypred. We will do this by first finding our new slope (b1), then finding our new intercept (b0).

In [4]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: 7756.425617967361
ypred new intercept: -2256.360580044546


The intercept in this model is negative but that is because it is the value expected when all independent variables are set to 0. It is a predicted value which is why it is able to be negative, however, if our x (carat) never equals 0 in the data set then this really has no essential meaning.

The slope for this model means for each carat we are esentially expecting an increase of $7756.43 in the price.

This calculations we use to find r-squared. When find r-squared it is important that we also find the total sum of squares and the total sum of squares of residuals in order to make sure we have all the components necessary to perform the r-squared equation.

In [5]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.8493305264354793


Results of the Linear Regression for Carat vs Price:

What an r-squared value of .849 means is that 84.9% of the variation in the output variable can be explained by the input variable. In other words,the there is a strong positive relationship between the carat of a diamond and the price of a diamond, which can be attributed due to having an r-squared value of .849. This also means that our original hypothesis was indeed correct that there is a relationship between the carat of a diamond and the price of a diamond. This makes sense because as I may have mentioned earlier, the carat of a diamond does generally tend to be more expensive the higher the value of the carat is.

Next up I would like to perform a correlation on both the variables price of a diamond and the carat of a diamond. A correlation is the measure of how closely related two variables are to each other. Specifically we are going to be looking at both Person's and Spearman's correlation. Pearson's correlation is the strength of the linear relationship between two variables, so somewhat similar to regression. Spearman's correlation measures the strength and direction of monotonic (less-restrictive) association between two variables.

In [6]:
corr, _ = spearmanr(new_x, new_y)
corr, _ = pearsonr(new_x, new_y)

print('Pearsons Correlation:  ', corr)
print('Spearmans Correlation: ', corr)

Pearsons Correlation:   0.9215913011934773
Spearmans Correlation:  0.9215913011934773


Results of the Correlation Testing for Carat vs Price:

As we see from the results for both methods we get a correlation of .922. This is a very strong correlation, thus indicating that there is a strong positive relationship between the two variables carat of a diamond and the price of a diamond. Which means that as the carat of a diamond increases the price of the diamond increases as well. This result gives us more confirmation with the results of our previous linear regression test.

Hypothesis for Carat vs Length, Width, and Depth: 

I believe that there will be a positive relationship between the carat of a diamond and the length of a diamond, since diamonds of high carat tend to have more value so I'm thinking they may bigger in size as well. Which is why I am keeping this hypothesis for length, width, and depth as well.

In [7]:
x = data['carat'].values
y = data['x'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) != 'nan':
        new_y.append(item)
        
print('The length of this dataset is:', len(new_x))

The length of this dataset is: 53940


In [8]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: 2.307587511637267
ypred new intercept: 3.8898414144975324


The intercept in this model means the predicted value of y when the carat is 0 would be 3.88mm in length. However, once again the data set has no carat of 0 in it so this would have no essential meaning it is simply a predicted value.

The slope for this model means for each carat we are esentially expecting an increase of 2.3mm for the length.

In [9]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.9508087509951948


Results of the Linear Regression for Carat vs Length:

What an r-squared value of .951 means is that 95.1% of the variation in the output variable can be explained by the input variable. In other words,the there is a strong positive relationship between the carat of a diamond and the length of a diamond. This makes sense because as I may have mentioned earlier, the carat of a diamond does generally tend to be more expensive the higher the value of the carat is, so it makes sense that the bigger diamond would have a higher carat value.

In [10]:
x = data['carat'].values
y = data['y'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) != 'nan':
        new_y.append(item)
        
print('The length of this dataset is:', len(new_x))

The length of this dataset is: 53940


In [11]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: 2.293183835732053
ypred new intercept: 3.904703423065543


The intercept in this model means the predicted value of y when the carat is 0 would be 3.9 mm in width. However, once again the data set has no carat of 0 in it so this would have no essential meaning it is simply a predicted value.

The slope for this model means for each carat we are esentially expecting an increase of 2.3mm for the width.

In [12]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.905775144094108


Results of the Linear Regression for Carat vs Width:

What an r-squared value of .906 means is that 90.6% of the variation in the output variable can be explained by the input variable. In other words,the there is a strong positive relationship between the carat of a diamond and the width of a diamond. This makes sense because as I may have mentioned earlier, the carat of a diamond does generally tend to be more expensive the higher the value of the carat is, so it makes sense that the bigger diamond would have a higher carat value.

In [13]:
x = data['carat'].values
y = data['z'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) != 'nan':
        new_y.append(item)
        
print('The length of this dataset is:', len(new_x))

The length of this dataset is: 53940


In [14]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: 1.4193848418146597
ypred new intercept: 2.4061501954669016


The intercept in this model means the predicted value of y when the carat is 0 would be 3.9 mm in width. However, once again the data set has no carat of 0 in it so this would have no essential meaning it is simply a predicted value.

The slope for this model means for each carat we are esentially expecting an increase of 2.3mm for the width.

In [15]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.908947497413759


Results of the Linear Regression for Carat vs Depth:

What an r-squared value of .909 means is that 90.9% of the variation in the output variable can be explained by the input variable. In other words,the there is a strong positive relationship between the carat of a diamond and the depth of a diamond. This makes sense because as I may have mentioned earlier, the carat of a diamond does generally tend to be more expensive the higher the value of the carat is, so it makes sense that the bigger diamond would have a higher carat value.

Hypothesis for Carat vs Clarity: 

I believe that there will be a positive relationship between the carat of a diamond and the clarity of a diamond, since diamonds of high carat tend to have more value I think that they would want to be more see through which may result in the diamonds being more expensive.

In [16]:
x = data['carat'].values
y = data['clarity'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) == 'I1':
        new_y.append(1)
    if str(item) == 'SI2':
        new_y.append(2)
    if str(item) == 'SI1':
        new_y.append(3)
    if str(item) == 'VS2':
        new_y.append(4)
    if str(item) == 'VS1':
        new_y.append(5)
    if str(item) == 'VVS2':
        new_y.append(6)
    if str(item) == 'VVS1':
        new_y.append(7)
    if str(item) == 'IF':
        new_y.append(8)
        
        
print('The length of this dataset is:', len(new_y))

The length of this dataset is: 53940


In [17]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: -1.2260815203644486
ypred new intercept: 5.029358830689814


The intercept in this model means the predicted value of y when the carat is 0 would be a clarity of 5 which is equivalent to VS1. However, once again the data set has no carat of 0 in it so this would have no essential meaning it is simply a predicted value. What I found interesting about this intercept value is that I would assume a carat of 0 would be a intercept of 0 since from my hypothesis I thought it would have the least clarity. However, according to these results it seems to not be the case therefore I already feel like there may not be much of a relationship between these two variables already.

The slope for this model means for each carat we are esentially expecting a decrease in the level of clarity by about 1 each time. Which doesn't make much sense to me once again since I expected the clarity level to increase as the carats of diamond moved up.

In [18]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.12449646687579252


Results of the Linear Regression for Carat vs Clarity:

What an r-squared value of .124 means that 12.4% of the variation in the output variable can be explained by the input variable. Which means there really isn't much of relationship between the carat of a diamond and the clarity of a diamond. This was a suprising result since it contradicts the hypothesis I made earlier, thus it seems there really isn't much of relationship between the carat of a diamond and the clarity of a diamond.